test/string.cpp

```diff
@@ -2290,6 +2188,129 @@ class string_test
     void
     testReplace()
     {
+        test_vectors const t;
+
+        // replace(std::size_t, std::size_t, string_view)
+        {
+            // outside, shrink
+            fail_loop([&](storage_ptr const& sp)
+            {
+                std::string s1(t.v2.data(), t.v2.size());
+                string s2(t.v2, sp);
+                BOOST_TEST(s2.replace(0, 4, t.v2.substr(4, 2)) == s1.replace(0, 4, t.v2.data() + 4, 2));
+            });
+
+            // outside, grow
+            fail_loop([&](storage_ptr const& sp)
+            {
+                std::string s1(t.v2.data(), t.v2.size());
+                string s2(t.v2, sp);
+                BOOST_TEST(s2.replace(0, 1, t.v2.substr(0)) == s1.replace(0, 1, t.v2.data(), t.v2.size()));
+            });
+
+            // outside, same
+            fail_loop([&](storage_ptr const& sp)
+            {
+                std::string s1(t.v2.data(), t.v2.size());
+                string s2(t.v2, sp);
+                BOOST_TEST(s2.replace(0, 2, t.v2.substr(0, 2)) == s1.replace(0, 2, t.v2.data(), 2));
+            });
+
+            // inside, shrink
+            fail_loop([&](storage_ptr const& sp)
+            {
+                std::string s1(t.v2.data(), t.v2.size());
+                string s2(t.v2, sp);
+                BOOST_TEST(s2.replace(1, 4, s2.substr(4, 2)) == s1.replace(1, 4, s1.data() + 4, 2));
+            });
+
+            // inside, grow
+            fail_loop([&](storage_ptr const& sp)
+            {
+                std::string s1(t.v2.data(), t.v2.size());
+                string s2(t.v2, sp);
+                BOOST_TEST(s2.replace(1, 1, s2.substr(0)) == s1.replace(1, 1, s1.data(), s1.size()));
+            });
+
+            // inside, same
+            fail_loop([&](storage_ptr const& sp)
+            {
+                std::string s1(t.v2.data(), t.v2.size());
+                string s2(t.v2, sp);
+                BOOST_TEST(s2.replace(1, 2, s2.subview(0, 2)) == s1.replace(1, 2, s1.data(), 2));
+            });
+        }
+
+        // replace(const_iterator, const_iterator, string_view)
+        {
+            // outside, shrink
+            fail_loop([&](storage_ptr const& sp)
+            {
+                std::string s1(t.v2.data(), t.v2.size());
+                string s2(t.v2, sp);
+                BOOST_TEST(s2.replace(s2.begin(), s2.begin() + 4, t.v2.substr(4, 2)) == s1.replace(0, 4, t.v2.data() + 4, 2));
+            });
+
+            // outside, grow
+            fail_loop([&](storage_ptr const& sp)
+            {
+                std::string s1(t.v2.data(), t.v2.size());
+                string s2(t.v2, sp);
+                BOOST_TEST(s2.replace(s2.begin(), s2.begin() + 1, t.v2.substr(0)) == s1.replace(0, 1, t.v2.data(), t.v2.size()));
```

Comment:
This is hard to read

Reply:
Reformatted:
```
BOOST_TEST(
    s2.replace(
        s2.begin(),
        s2.begin() + 1,
        t.v2.substr(0)) ==
    s1.replace(
        0, 1,
        t.v2.data(),
        t.v2.size()));
```
Now that I can read it... we are using the iterator overload of `json::string` but the index overload of `std::string`? Why?

Reply:
Reformatted in new commit.

> Now that I can read it... we are using the iterator overload of `json::string` but the index overload of `std::string`? Why?

They perform the same operation, it was just easier to copy the index based tests.